program-runtime/src/loaded_programs.rs

```diff
@@ -302,6 +337,8 @@ impl LoadedPrograms {
             second_level.reverse();
             !second_level.is_empty()
         });
+
+        self.remove_expired_entries(new_root);
```

Comment:
Instead of calling `remove_programs_with_no_entries()` inside `remove_expired_entries()` I would move it here so that it becomes more obvious that this cleans both, expired second level and then empty first level entries.